File: contracts/api/clients/jakarta/src/main/java/app/epistola/client/jakarta/collect/ResultCollector.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.jakarta.collect;

import app.epistola.client.jakarta.EpistolaJson;
import app.epistola.client.jakarta.model.CollectMeta;
import app.epistola.client.jakarta.model.CollectRequest;
import app.epistola.client.jakarta.model.GenerationResult;
import app.epistola.client.jakarta.model.PartitionAssignment;
import app.epistola.protocol.Compression;
import app.epistola.protocol.PartitionRouting;
import app.epistola.protocol.PollBackoff;
import jakarta.ws.rs.core.Response;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

/**
 * Collects generation results from {@code /generation/collect} with NDJSON streaming,
 * compression and adaptive polling.
 *
 * <p>Results are processed one at a time — the response is never held in memory. Results from this
 * node come first; orphaned results from nodes that stopped polling follow.
 *
 * <p>What it handles:
 * <ul>
 *   <li>NDJSON streaming (constant memory)</li>
 *   <li>Compression: gzip built in, lz4/zstd used when those libraries are on the classpath</li>
 *   <li>Adaptive polling — immediate while {@code hasMore}, exponential backoff when idle</li>
 *   <li>Sequence-based acknowledgement: safe across restarts, and a batch whose handler threw
 *       is redelivered rather than lost</li>
 *   <li>Partition-aware routing-key helpers</li>
 *   <li>Observability through {@link MetricsListener}</li>
 *   <li>A thread-safe {@link #collectOnce()} for scheduling the polls yourself</li>
 * </ul>
 *
 * <pre>{@code
 * ResultCollector collector = ResultCollector.builder()
 *     .collectApi(collectApi)
 *     .tenantId("acme-corp")
 *     .handler(result -> {
 *         switch (result.getStatus()) {
 *             case COMPLETED -> downloadAndProcess(result.getDocumentId(), result.getCorrelationId());
 *             case FAILED -> logFailure(result.getCorrelationId(), result.getError());
 *         }
 *     })
 *     .build();
 *
 * collector.start();  // blocks, runs the adaptive poll loop
 * }</pre>
 *
 * <p>In a Jakarta EE application, run {@link #start()} on a managed thread
 * ({@code ManagedExecutorService}) or drive {@link #collectOnce()} from a
 * {@code @Schedule} timer — never on an unmanaged {@code new Thread(...)}.
 */
public final class ResultCollector {

    private final GenerationCollectApi collectApi;
    private final String tenantId;
    private final int batchSize;
    private final Duration minInterval;
    private final Duration maxInterval;
    private final Duration kickInterval;
    private final Consumer<GenerationResult> handler;
    private final Consumer<Exception> errorHandler;
    private final MetricsListener metricsListener;
    private final boolean registerShutdownHook;

    /**
     * How many prefixes {@link #routingKeyToMe(String)} tries before giving up. Reaching this
     * means every one of a thousand hashes missed every partition this node owns.
     */
    private static final int MAX_ROUTING_KEY_ATTEMPTS = 1000;

    // The polling policy — shared with the Kotlin client, because the collapse-to-zero bug lived
    // in this arithmetic in every client.
    private final PollBackoff backoff;

    private final AtomicBoolean running = new AtomicBoolean(false);
    private final ReentrantLock pollLock = new ReentrantLock();

    // Wakeable sleep — replaces Thread.sleep so kick() and stop() can cut an in-progress wait
    // short. Capacity 1 is intentional: rapid repeat calls collapse into a single wake-up.
    private final LinkedBlockingQueue<Boolean> wakeUp = new LinkedBlockingQueue<>(1);

    private volatile long currentInterval;
    private volatile PartitionAssignment partitionAssignment;
    private Long lastAcknowledgedSequence;
    private Thread shutdownHook;

    private ResultCollector(Builder builder) {
        this.collectApi = builder.collectApi;
        this.tenantId = builder.tenantId;
        this.batchSize = builder.batchSize;
        this.minInterval = builder.minInterval;
        this.maxInterval = builder.maxInterval;
        this.kickInterval = builder.kickInterval;
        this.handler = builder.handler;
        this.errorHandler = builder.errorHandler;
        this.metricsListener = builder.metricsListener;
        this.registerShutdownHook = builder.registerShutdownHook;
        this.backoff = PollBackoff.of(
                builder.minInterval.toMillis(), builder.maxInterval.toMillis(), builder.backoffMultiplier);
        this.currentInterval = builder.minInterval.toMillis();
    }

    public static Builder builder() {
        return new Builder();
    }

    /** The outcome of a single collection. */
    public static final class CollectResult {

        private final int count;
        private final boolean hasMore;

        CollectResult(int count, boolean hasMore) {
            this.count = count;
            this.hasMore = hasMore;
        }

        /** How many results the handler was called for. */
        public int getCount() {
            return count;
        }

        /** True when the server still has results queued for this node. */
        public boolean isHasMore() {
            return hasMore;
        }
    }

    /** Observability hooks. Implementations must not block the poll loop. */
    public interface MetricsListener {

        /** Called after each poll completes, successfully or not. */
        void onPoll(int count, boolean hasMore, long durationMs, Exception error);

        /** Called when the server hands this node a different set of partitions. */
        void onPartitionChange(PartitionAssignment oldAssignment, PartitionAssignment newAssignment);
    }

    /** The current partition assignment, refreshed from the {@code _meta} line of every poll. */
    public PartitionAssignment getPartitionAssignment() {
        return partitionAssignment;
    }

    // --- Partition routing helpers ---

    /** The current assignment as the shared router sees it, or null while it is unknown. */
    private PartitionRouting routing() {
        PartitionAssignment assignment = partitionAssignment;
        return assignment == null ? null : PartitionRouting.of(assignment.getTotal(), assignment.getMine());
    }

    /**
     * The partition a routing key lands on, using the server's hash (murmur3 x86 32-bit, seed 0).
     * Returns {@code null} until the first poll has reported an assignment.
     */
    public Integer partitionFor(String routingKey) {
        PartitionRouting routing = routing();
        return routing == null ? null : routing.partitionFor(routingKey);
    }

    /** True when {@code routingKey} would land on one of this node's partitions. */
    public boolean isMyPartition(String routingKey) {
        PartitionRouting routing = routing();
        return routing != null && routing.isMine(routingKey);
    }

    /**
     * A routing key that targets one of this node's partitions — useful to make a submission's
     * result come back to the node that made it. Returns {@code null} until the first poll has
     * reported an assignment.
     *
     * <p>The prefix is what the server hashes, so a rewritten key is a different key: pass the
     * value returned here as the request's {@code routingKey}, and expect it back on the result.
     */
    public String routingKeyToMe(String key) {
        PartitionRouting routing = routing();
        return routing == null ? null : routing.routingKeyToMe(key);
    }

    // --- Poll loop ---

    /** Runs the adaptive poll loop, blocking the calling thread until {@link #stop()}. */
    public void start() {
        running.set(true);
        currentInterval = minInterval.toMillis();

        if (registerShutdownHook) {
            shutdownHook = new Thread(this::stop, "ResultCollector-shutdown");
            Runtime.getRuntime().addShutdownHook(shutdownHook);
        }

        try {
            while (running.get()) {
                try {
                    CollectResult result = collectOnce();
                    if (!running.get()) {
                        break;
                    }
                    if (result.isHasMore()) {
                        currentInterval = backoff.afterHasMore();
                    } else if (result.getCount() > 0) {
                        currentInterval = backoff.afterResults();
                    } else {
                        currentInterval = backoff.afterIdlePoll(currentInterval);
                    }
                    sleepInterruptibly(currentInterval);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                } catch (Exception e) {
                    if (errorHandler != null) {
                        errorHandler.accept(e);
                    }
                    long jitter = ThreadLocalRandom.current().nextLong(currentInterval / 2 + 1);
                    currentInterval = backoff.afterIdlePoll(currentInterval);
                    try {
                        sleepInterruptibly(currentInterval + jitter);
                    } catch (InterruptedException interrupted) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }
            }
        } finally {
            removeShutdownHook();
        }
    }

    /**
     * Hints that a result is expected soon — typically called straight after a submission, so the
     * next poll happens within the kick interval instead of waiting out the full backoff.
     *
     * <p>Threshold-guarded: while the collector is already polling at or below the kick interval
     * the next poll is imminent anyway, and waking early would only burn a poll. So this does
     * nothing during active periods and matters only once the collector has backed off into idle.
     *
     * <p>Safe to call from any thread, any number of times.
     */
    public void kick() {
        long shortened = backoff.afterKick(currentInterval, kickInterval.toMillis());
        if (shortened != currentInterval) {
            currentInterval = shortened;
            wakeUp.offer(Boolean.TRUE);
        }
    }

    /** Signals the poll loop to stop once the current collection completes. */
    public void stop() {
        running.set(false);
        // Wake the loop so shutdown is not delayed by an in-progress backoff.
        wakeUp.offer(Boolean.TRUE);
    }

    /**
     * Performs a single collection. Thread-safe: concurrent calls are serialized so a result is
     * never delivered twice.
     *
     * <p>Streams the NDJSON response line by line, calling the handler for each result. If the
     * handler throws, the acknowledgement cursor is not advanced past the failing result and the
     * rest of the batch is redelivered on the next call.
     */
    public CollectResult collectOnce() {
        pollLock.lock();
        try {
            long startTime = System.currentTimeMillis();
            try {
                CollectResult result = poll();
                if (metricsListener != null) {
                    metricsListener.onPoll(
                            result.getCount(), result.isHasMore(), System.currentTimeMillis() - startTime, null);
                }
                return result;
            } catch (RuntimeException e) {
                if (metricsListener != null) {
                    metricsListener.onPoll(0, false, System.currentTimeMillis() - startTime, e);
                }
                throw e;
            }
        } finally {
            pollLock.unlock();
        }
    }

    private CollectResult poll() {
        CollectRequest request = new CollectRequest().limit(batchSize);
        if (lastAcknowledgedSequence != null) {
            request.setAcknowledgeUpTo(lastAcknowledgedSequence);
        }

        try (Response response =
                collectApi.collectGenerationResults(tenantId, Compression.acceptEncoding(), request)) {
            InputStream body = response.readEntity(InputStream.class);
            if (body == null) {
                return new CollectResult(0, false);
            }

            int count = 0;
            boolean hasMore = false;
            Long lastSequenceInBatch = null;

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(Compression.decompress(body), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.isBlank()) {
                        continue;
                    }
                    GenerationResult result = EpistolaJson.jsonb().fromJson(line, GenerationResult.class);
                    if (result.getSequence() == null) {
                        // Only the trailing _meta line lacks a sequence; re-read it as such.
                        hasMore = applyMeta(line);
                        break;
                    }
                    handler.accept(result);
                    lastSequenceInBatch = result.getSequence();
                    count++;
                }
            } catch (IOException e) {
                throw new UncheckedIOException("Failed to read the collect response stream", e);
            }

            // Only reached when every line was handled: a throwing handler leaves the cursor
            // where it was, so the whole batch is redelivered rather than silently dropped.
            if (lastSequenceInBatch != null) {
                lastAcknowledgedSequence = lastSequenceInBatch;
            }

            return new CollectResult(count, hasMore);
        }
    }

    private boolean applyMeta(String line) {
        CollectMeta meta = EpistolaJson.jsonb().fromJson(line, CollectMeta.class);
        PartitionAssignment assignment = meta.getPartitions();
        if (assignment != null && !Objects.equals(partitionAssignment, assignment)) {
            PartitionAssignment previous = partitionAssignment;
            partitionAssignment = assignment;
            if (metricsListener != null) {
                metricsListener.onPartitionChange(previous, assignment);
            }
        }
        return Boolean.TRUE.equals(meta.getHasMore());
    }

    /**
     * Wakeable replacement for {@code Thread.sleep}: returns early when {@link #kick()} or
     * {@link #stop()} offers a token, and drains any extra tokens afterwards so a burst of calls
     * cannot shorten the next wait too.
     */
    private void sleepInterruptibly(long durationMs) throws InterruptedException {
        if (durationMs <= 0) {
            wakeUp.clear();
            return;
        }
        wakeUp.poll(durationMs, TimeUnit.MILLISECONDS);
        wakeUp.clear();
    }

    private void removeShutdownHook() {
        if (shutdownHook == null) {
            return;
        }
        try {
            Runtime.getRuntime().removeShutdownHook(shutdownHook);
        } catch (IllegalStateException e) {
            // The JVM is already shutting down; the hook is running or has run.
        }
        shutdownHook = null;
    }

    /** Builds a {@link ResultCollector}. {@code collectApi}, {@code tenantId} and {@code handler} are required. */
    public static final class Builder {

        private GenerationCollectApi collectApi;
        private String tenantId;
        private int batchSize = 100;
        private Duration minInterval = Duration.ofSeconds(1);
        private Duration maxInterval = Duration.ofSeconds(30);
        private Duration kickInterval = Duration.ofSeconds(3);
        private double backoffMultiplier = 3.0;
        private Consumer<GenerationResult> handler;
        private Consumer<Exception> errorHandler;
        private MetricsListener metricsListener;
        private boolean registerShutdownHook = true;

        /** The collect endpoint — inject {@code @RestClient GenerationCollectApi} or build one. */
        public Builder collectApi(GenerationCollectApi collectApi) {
            this.collectApi = collectApi;
            return this;
        }

        public Builder tenantId(String tenantId) {
            this.tenantId = tenantId;
            return this;
        }

        /** Maximum results per collection (default: 100). */
        public Builder batchSize(int batchSize) {
            if (batchSize < 1 || batchSize > 10000) {
                throw new IllegalArgumentException("batchSize must be between 1 and 10000");
            }
            this.batchSize = batchSize;
            return this;
        }

        /** Minimum poll interval while results are flowing (default: 1s). */
        public Builder minInterval(Duration minInterval) {
            this.minInterval = requirePositive(minInterval, "minInterval");
            return this;
        }

        /** Maximum poll interval when idle (default: 30s). */
        public Builder maxInterval(Duration maxInterval) {
            this.maxInterval = requirePositive(maxInterval, "maxInterval");
            return this;
        }

        /**
         * Wait used by {@link #kick()} to override the current backoff (default: 3s). Short enough
         * to feel responsive, long enough that the server has had a chance to produce the expected
         * result before the poll asks for it.
         */
        public Builder kickInterval(Duration kickInterval) {
            this.kickInterval = requirePositive(kickInterval, "kickInterval");
            return this;
        }

        /**
         * Exponential backoff multiplier applied on each empty poll (default: 3.0). With the
         * default 1s minimum and 30s maximum that is 1s → 3s → 9s → 27s → 30s. Higher multipliers
         * reach the maximum sooner and cut idle poll volume; {@link #kick()} is the safety net that
         * gets back to fast polling when a result is expected.
         */
        public Builder backoffMultiplier(double backoffMultiplier) {
            if (backoffMultiplier <= 1.0) {
                throw new IllegalArgumentException("backoffMultiplier must be > 1.0");
            }
            this.backoffMultiplier = backoffMultiplier;
            return this;
        }

        /** Called for each result as it streams in. Throwing leaves the batch unacknowledged from that point. */
        public Builder handler(Consumer<GenerationResult> handler) {
            this.handler = handler;
            return this;
        }

        /** Optional handler for collection failures; the loop backs off and retries either way. */
        public Builder errorHandler(Consumer<Exception> errorHandler) {
            this.errorHandler = errorHandler;
            return this;
        }

        /** Optional metrics listener. */
        public Builder metricsListener(MetricsListener metricsListener) {
            this.metricsListener = metricsListener;
            return this;
        }

        /**
         * Registers a JVM shutdown hook that stops polling gracefully (default: true). Turn it off
         * inside an application server, where the container's lifecycle callbacks own shutdown.
         */
        public Builder registerShutdownHook(boolean registerShutdownHook) {
            this.registerShutdownHook = registerShutdownHook;
            return this;
        }

        public ResultCollector build() {
            if (collectApi == null) {
                throw new IllegalStateException("collectApi is required");
            }
            if (tenantId == null) {
                throw new IllegalStateException("tenantId is required");
            }
            if (handler == null) {
                throw new IllegalStateException("handler is required");
            }
            return new ResultCollector(this);
        }

        private static Duration requirePositive(Duration value, String name) {
            if (value == null || value.isNegative() || value.isZero()) {
                throw new IllegalArgumentException(name + " must be positive");
            }
            return value;
        }
    }
}

File: client-kotlin-spring-restclient/client/src/main/kotlin/app/epistola/client/collect/ResultCollector.kt
package app.epistola.client.collect

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.web.client.RestClient
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.time.Duration
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.locks.ReentrantLock
import java.util.zip.GZIPInputStream
import kotlin.concurrent.withLock

/**
 * Collects generation results via the `/generation/collect` endpoint with
 * NDJSON streaming, compression, and adaptive polling.
 *
 * Results are processed one at a time — the response is never loaded into memory.
 * Results from your node are returned first; orphaned results from dead nodes follow.
 *
 * Features:
 * - NDJSON streaming (constant memory)
 * - Compression: gzip built-in, lz4/zstd auto-detected if libraries on classpath
 * - Adaptive polling (immediate on hasMore, exponential backoff when idle)
 * - Sequence-based acknowledgment (safe on restart, partial ack)
 * - Partition-aware routing key helpers
 * - Metrics via [MetricsListener]
 * - Thread-safe [collectOnce] for custom scheduling
 * - Shutdown hook for graceful stop
 *
 * Example:
 * ```
 * val collector = ResultCollector.builder()
 *     .restClient(restClient)
 *     .tenantId("acme-corp")
 *     .handler { result ->
 *         when (result.status) {
 *             "COMPLETED" -> downloadAndProcess(result.documentId, result.correlationId)
 *             "FAILED" -> logFailure(result.correlationId, result.error)
 *         }
 *     }
 *     .build()
 *
 * collector.start()  // blocks, runs adaptive poll loop
 * ```
 */
class ResultCollector private constructor(
    private val restClient: RestClient,
    private val tenantId: String,
    private val batchSize: Int,
    private val minInterval: Duration,
    private val maxInterval: Duration,
    private val handler: (GenerationResult) -> Unit,
    private val errorHandler: ((Exception) -> Unit)?,
    private val metricsListener: MetricsListener?,
    private val objectMapper: ObjectMapper,
    private val registerShutdownHook: Boolean,
) {
    private val running = AtomicBoolean(false)
    private var currentInterval: Long = minInterval.toMillis()
    private var lastAcknowledgedSequence: Long? = null
    private val pollLock = ReentrantLock()
    private var shutdownHook: Thread? = null

    /** Current partition assignment, updated on each poll from the _meta line. */
    @Volatile
    var partitionAssignment: PartitionAssignment? = null
        private set

    companion object {
        private val NDJSON = MediaType.parseMediaType("application/vnd.epistola.v1+ndjson")
        private val EPISTOLA_JSON = MediaType.parseMediaType("application/vnd.epistola.v1+json")

        // Optional decompressors — loaded via reflection to avoid hard dependencies
        private val lz4Decompressor: ((InputStream) -> InputStream)? = tryLoadLz4()
        private val zstdDecompressor: ((InputStream) -> InputStream)? = tryLoadZstd()

        @Suppress("ktlint:standard:function-signature")
        private fun tryLoadLz4(): ((InputStream) -> InputStream)? =
            try {
                val ctor = Class.forName("net.jpountz.lz4.LZ4FrameInputStream")
                    .getConstructor(InputStream::class.java)
                val fn: (InputStream) -> InputStream = { input -> ctor.newInstance(input) as InputStream }
                fn
            } catch (_: Exception) {
                null
            }

        @Suppress("ktlint:standard:function-signature")
        private fun tryLoadZstd(): ((InputStream) -> InputStream)? =
            try {
                val ctor = Class.forName("com.github.luben.zstd.ZstdInputStream")
                    .getConstructor(InputStream::class.java)
                val fn: (InputStream) -> InputStream = { input -> ctor.newInstance(input) as InputStream }
                fn
            } catch (_: Exception) {
                null
            }

        fun builder(): Builder = Builder()
    }

    /** A completed or failed generation result. */
    data class GenerationResult(
        val sequence: Long,
        val requestId: String,
        val batchId: String?,
        val status: String,
        val documentId: String?,
        val correlationId: String?,
        val routingKey: String?,
        val templateId: String?,
        val variantId: String?,
        val versionId: Int?,
        val filename: String?,
        val contentType: String?,
        val sizeBytes: Long?,
        val error: String?,
        val completedAt: String?,
    )

    data class CollectResult(
        val count: Int,
        val hasMore: Boolean,
    )

    /** Partition assignment info from the server. */
    data class PartitionAssignment(
        val total: Int,
        val mine: List<Int>,
        val hash: String,
    )

    /** Callback interface for observability. */
    interface MetricsListener {
        /** Called after each poll completes. */
        fun onPoll(count: Int, hasMore: Boolean, durationMs: Long, error: Exception? = null)

        /** Called when partition assignment changes. */
        fun onPartitionChange(oldAssignment: PartitionAssignment?, newAssignment: PartitionAssignment)
    }

    // --- Partition routing helpers ---

    /**
     * Compute the partition number for a given routing key using the same
     * hash algorithm as the server (murmur3 x86 32-bit, seed 0).
     *
     * Returns null if the partition assignment is not yet known (call after first poll).
     */
    fun partitionFor(routingKey: String): Int? {
        val assignment = partitionAssignment ?: return null
        val hash = murmur3x86_32(routingKey.toByteArray(Charsets.UTF_8), 0)
        return (hash and 0x7FFFFFFF) % assignment.total
    }

    /**
     * Check if a routing key would land on one of this node's partitions.
     */
    fun isMyPartition(routingKey: String): Boolean {
        val partition = partitionFor(routingKey) ?: return false
        return partition in (partitionAssignment?.mine ?: emptyList())
    }

    /**
     * Find a routing key prefix that targets one of this node's partitions.
     * Appends a partition number prefix to the given key to ensure it routes to this node.
     *
     * Returns the original key if it already routes here, or a prefixed key if not.
     * Returns null if partition assignment is not yet known.
     */
    fun routingKeyToMe(key: String): String? {
        val assignment = partitionAssignment ?: return null
        if (isMyPartition(key)) return key
        for (p in assignment.mine) {
            val candidate = "$p:$key"
            if (isMyPartition(candidate)) return candidate
        }
        return "${assignment.mine.first()}:$key"
    }

    // --- Poll loop ---

    /**
     * Starts the adaptive poll loop. Blocks the current thread until [stop] is called.
     */
    fun start() {
        running.set(true)
        currentInterval = minInterval.toMillis()

        if (registerShutdownHook) {
            shutdownHook = Thread({ stop() }, "ResultCollector-shutdown").also {
                Runtime.getRuntime().addShutdownHook(it)
            }
        }

        try {
            while (running.get()) {
                try {
                    val result = collectOnce()
                    if (!running.get()) break

                    currentInterval = when {
                        result.hasMore -> 0
                        result.count > 0 -> minInterval.toMillis()
                        else -> (currentInterval * 2).coerceAtMost(maxInterval.toMillis())
                    }

                    if (currentInterval > 0) {
                        Thread.sleep(currentInterval)
                    }
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                    break
                } catch (e: Exception) {
                    errorHandler?.invoke(e)
                    val jitter = ThreadLocalRandom.current().nextLong(currentInterval / 2 + 1)
                    currentInterval = (currentInterval * 2).coerceAtMost(maxInterval.toMillis())
                    try {
                        Thread.sleep(currentInterval + jitter)
                    } catch (_: InterruptedException) {
                        Thread.currentThread().interrupt()
                        break
                    }
                }
            }
        } finally {
            removeShutdownHook()
        }
    }

    /** Signals the poll loop to stop after the current collection completes. */
    fun stop() {
        running.set(false)
    }

    /**
     * Performs a single collection call. Thread-safe — concurrent calls are serialized
     * via a lock to prevent duplicate delivery.
     *
     * Streams the NDJSON response line by line, calling the handler for each result.
     * If the handler throws, the sequence is not advanced and the batch will be
     * redelivered on the next call.
     */
    fun collectOnce(): CollectResult = pollLock.withLock {
        val startTime = System.currentTimeMillis()
        var error: Exception? = null

        try {
            val requestBody = buildString {
                append("{")
                if (lastAcknowledgedSequence != null) {
                    append("\"acknowledgeUpTo\":").append(lastAcknowledgedSequence)
                    append(",")
                }
                append("\"limit\":").append(batchSize)
                append("}")
            }

            val result = restClient.method(HttpMethod.POST)
                .uri("/tenants/{tenantId}/generation/collect", tenantId)
                .contentType(EPISTOLA_JSON)
                .accept(NDJSON)
                .header("Accept-Encoding", supportedEncodings())
                .body(requestBody)
                .exchange { _, response ->
                    val stream = decompressIfNeeded(
                        response.body,
                        response.headers.getFirst("Content-Encoding"),
                    )
                    val reader = BufferedReader(InputStreamReader(stream, Charsets.UTF_8))

                    var count = 0
                    var hasMore = false
                    var lastSequenceInBatch: Long? = null

                    reader.useLines { lines ->
                        for (line in lines) {
                            if (line.isBlank()) continue
                            val node = objectMapper.readTree(line)

                            if (node.has("_meta") && node["_meta"].asBoolean()) {
                                hasMore = node["hasMore"]?.asBoolean() ?: false
                                updatePartitionAssignment(node)
                                break
                            }

                            val parsed = parseResult(node)
                            handler(parsed)
                            lastSequenceInBatch = parsed.sequence
                            count++
                        }
                    }

                    if (lastSequenceInBatch != null) {
                        lastAcknowledgedSequence = lastSequenceInBatch
                    }

                    val collectResult = CollectResult(count = count, hasMore = hasMore)
                    metricsListener?.onPoll(count, hasMore, System.currentTimeMillis() - startTime)
                    collectResult
                }

            return result ?: CollectResult(count = 0, hasMore = false)
        } catch (e: Exception) {
            error = e
            metricsListener?.onPoll(0, false, System.currentTimeMillis() - startTime, e)
            throw e
        }
    }

    private fun updatePartitionAssignment(metaNode: com.fasterxml.jackson.databind.JsonNode) {
        val partitionsNode = metaNode["partitions"] ?: return
        val newAssignment = PartitionAssignment(
            total = partitionsNode["total"]?.asInt() ?: return,
            mine = partitionsNode["mine"]?.map { it.asInt() } ?: return,
            hash = partitionsNode["hash"]?.asText() ?: "murmur3",
        )
        val old = partitionAssignment
        if (old != newAssignment) {
            partitionAssignment = newAssignment
            metricsListener?.onPartitionChange(old, newAssignment)
        }
    }

    private fun parseResult(node: com.fasterxml.jackson.databind.JsonNode) = GenerationResult(
        sequence = node["sequence"].asLong(),
        requestId = node["requestId"].asText(),
        batchId = node["batchId"]?.asText(),
        status = node["status"].asText(),
        documentId = node["documentId"]?.asText(),
        correlationId = node["correlationId"]?.asText(),
        routingKey = node["routingKey"]?.asText(),
        templateId = node["templateId"]?.asText(),
        variantId = node["variantId"]?.asText(),
        versionId = node["versionId"]?.asInt(),
        filename = node["filename"]?.asText(),
        contentType = node["contentType"]?.asText(),
        sizeBytes = node["sizeBytes"]?.asLong(),
        error = node["error"]?.asText(),
        completedAt = node["completedAt"]?.asText(),
    )

    private fun supportedEncodings(): String = buildList {
        if (lz4Decompressor != null) add("lz4")
        if (zstdDecompressor != null) add("zstd")
        add("gzip")
    }.joinToString(", ")

    private fun decompressIfNeeded(input: InputStream, encoding: String?): InputStream = when (encoding) {
        "gzip" -> GZIPInputStream(input)
        "lz4" -> lz4Decompressor?.invoke(input)
            ?: throw IllegalStateException("Server sent lz4 but net.jpountz.lz4:lz4-java is not on classpath")
        "zstd" -> zstdDecompressor?.invoke(input)
            ?: throw IllegalStateException("Server sent zstd but com.github.luben:zstd-jni is not on classpath")
        else -> input
    }

    private fun removeShutdownHook() {
        shutdownHook?.let {
            try {
                Runtime.getRuntime().removeShutdownHook(it)
            } catch (_: IllegalStateException) {
                // JVM is already shutting down
            }
        }
        shutdownHook = null
    }

    class Builder {
        private var restClient: RestClient? = null
        private var tenantId: String? = null
        private var batchSize: Int = 100
        private var minInterval: Duration = Duration.ofSeconds(1)
        private var maxInterval: Duration = Duration.ofSeconds(30)
        private var handler: ((GenerationResult) -> Unit)? = null
        private var errorHandler: ((Exception) -> Unit)? = null
        private var metricsListener: MetricsListener? = null
        private var objectMapper: ObjectMapper? = null
        private var registerShutdownHook: Boolean = true

        fun restClient(client: RestClient) = apply { this.restClient = client }
        fun tenantId(tenantId: String) = apply { this.tenantId = tenantId }

        /** Maximum results per collection (default: 100). */
        fun batchSize(size: Int) = apply {
            require(size in 1..10000) { "batchSize must be between 1 and 10000" }
            this.batchSize = size
        }

        /** Minimum poll interval when results are flowing (default: 1s). */
        fun minInterval(interval: Duration) = apply {
            require(!interval.isNegative && !interval.isZero) { "minInterval must be positive" }
            this.minInterval = interval
        }

        /** Maximum poll interval when idle (default: 30s). */
        fun maxInterval(interval: Duration) = apply {
            require(!interval.isNegative && !interval.isZero) { "maxInterval must be positive" }
            this.maxInterval = interval
        }

        /** Handler called for each result as it streams in. */
        fun handler(handler: (GenerationResult) -> Unit) = apply { this.handler = handler }

        /** Optional error handler for collection failures. */
        fun errorHandler(handler: (Exception) -> Unit) = apply { this.errorHandler = handler }

        /** Optional metrics listener for observability. */
        fun metricsListener(listener: MetricsListener) = apply { this.metricsListener = listener }

        /** Custom ObjectMapper (default: Jackson with JavaTimeModule). */
        fun objectMapper(mapper: ObjectMapper) = apply { this.objectMapper = mapper }

        /** Register a JVM shutdown hook to stop polling gracefully (default: true). */
        fun registerShutdownHook(register: Boolean) = apply { this.registerShutdownHook = register }

        @Suppress("ktlint:standard:function-signature")
        fun build(): ResultCollector =
            ResultCollector(
                restClient = requireNotNull(restClient) { "restClient is required" },
                tenantId = requireNotNull(tenantId) { "tenantId is required" },
                batchSize = batchSize,
                minInterval = minInterval,
                maxInterval = maxInterval,
                handler = requireNotNull(handler) { "handler is required" },
                errorHandler = errorHandler,
                metricsListener = metricsListener,
                objectMapper = objectMapper ?: ObjectMapper().apply {
                    registerModule(JavaTimeModule())
                },
                registerShutdownHook = registerShutdownHook,
            )
    }
}

/**
 * MurmurHash3 x86 32-bit with configurable seed.
 * Matches the server-side implementation (Guava `Hashing.murmur3_32_fixed(seed)`).
 */
@Suppress("ktlint:standard:function-naming")
internal fun murmur3x86_32(data: ByteArray, seed: Int): Int {
    val c1 = 0xcc9e2d51.toInt()
    val c2 = 0x1b873593
    var h1 = seed
    val len = data.size
    val nblocks = len / 4

    for (i in 0 until nblocks) {
        val idx = i * 4
        var k1 = (data[idx].toInt() and 0xFF) or
            ((data[idx + 1].toInt() and 0xFF) shl 8) or
            ((data[idx + 2].toInt() and 0xFF) shl 16) or
            ((data[idx + 3].toInt() and 0xFF) shl 24)

        k1 *= c1
        k1 = Integer.rotateLeft(k1, 15)
        k1 *= c2
        h1 = h1 xor k1
        h1 = Integer.rotateLeft(h1, 13)
        h1 = h1 * 5 + 0xe6546b64.toInt()
    }

    val tail = nblocks * 4
    var k1 = 0
    @Suppress("ktlint:standard:statement-wrapping")
    when (len and 3) {
        3 -> {
            k1 = k1 xor ((data[tail + 2].toInt() and 0xFF) shl 16)
            k1 = k1 xor ((data[tail + 1].toInt() and 0xFF) shl 8)
            k1 = k1 xor (data[tail].toInt() and 0xFF)
            k1 *= c1
            k1 = Integer.rotateLeft(k1, 15)
            k1 *= c2
            h1 = h1 xor k1
        }
        2 -> {
            k1 = k1 xor ((data[tail + 1].toInt() and 0xFF) shl 8)
            k1 = k1 xor (data[tail].toInt() and 0xFF)
            k1 *= c1
            k1 = Integer.rotateLeft(k1, 15)
            k1 *= c2
            h1 = h1 xor k1
        }
        1 -> {
            k1 = k1 xor (data[tail].toInt() and 0xFF)
            k1 *= c1
            k1 = Integer.rotateLeft(k1, 15)
            k1 *= c2
            h1 = h1 xor k1
        }
    }

    h1 = h1 xor len
    h1 = h1 xor (h1 ushr 16)
    h1 *= 0x85ebca6b.toInt()
    h1 = h1 xor (h1 ushr 13)
    h1 *= 0xc2b2ae35.toInt()
    h1 = h1 xor (h1 ushr 16)

    return h1
}
